File: src/main/java/fr/lewon/bot/runner/lifecycle/bot/BotLifeCycleOperation.java
package fr.lewon.bot.runner.lifecycle.bot;

import fr.lewon.bot.runner.lifecycle.ILifeCycleOperation;
import fr.lewon.bot.runner.lifecycle.Transition;

import static fr.lewon.bot.runner.lifecycle.bot.BotState.*;

public enum BotLifeCycleOperation implements ILifeCycleOperation<BotState> {

    START(new Transition<>(ACTIVE, PENDING, STOPPED)),
    STOP(new Transition<>(STOPPED, ACTIVE)),
    KILL(new Transition<>(KILLED, PENDING, ACTIVE, STOPPED));

    private Transition<BotState>[] possibleTransitions;

    BotLifeCycleOperation(Transition... possibleTransitions) {
        this.possibleTransitions = possibleTransitions;
    }

    @Override
    public Transition<BotState>[] getTransitions() {
        return this.possibleTransitions;
    }
}

File: src/main/kotlin/fr/lewon/bot/runner/lifecycle/bot/BotLifeCycleOperation.kt
package fr.lewon.bot.runner.lifecycle.bot

import fr.lewon.bot.runner.Bot
import fr.lewon.bot.runner.errors.InvalidOperationException

enum class BotLifeCycleOperation(private val operation: (Bot) -> Unit) {

    START({ it.start() }),
    STOP({ it.stop() });

    fun run(bot: Bot) {
        if (!bot.state.operations.contains(this)) {
            throw InvalidOperationException(name, bot.state.name)
        }
        val oldState = bot.state
        operation.invoke(bot)
        bot.logger.info("Bot transition applied : [$oldState] => [${bot.state}]")
    }

}